hack/helm/release/internal/helm/helm.go

```diff
@@ -0,0 +1,616 @@
+	// upload charts without dependencies
+	if err := uploadCharts(ctx, tempDir, conf.noDeps, conf.releaseConf); err != nil {
+		return err
+	}
+	if err := updateIndex(ctx, tempDir, conf); err != nil {
+		return err
+	}
+	// This retry is here because of caching in front of the Helm repository
+	// and the time it takes for a new release to show up in the repository.
+	// If the eck-stack chart depends on new version of any of the other
+	// eck-resources charts, and that new version is just released, then
+	// it will take ~ one hour for it to show up, so we will continue trying
+	// to get all dependencies of the helm charts, and upload them for 1 hour.
+	retry.Do(
+		func() error {
+			if err := updateHelmRepositories(); err != nil {
+				return err
+			}
+			// upload charts with dependencies
+			if err := uploadCharts(ctx, tempDir, conf.withDeps, conf.releaseConf); err != nil {
+				return err
+			}
+			if err := updateIndex(ctx, tempDir, conf); err != nil {
+				return err
+			}
+
+			return nil
+		},
+		retry.RetryIf(func(err error) bool {
+			if strings.Contains(err.Error(), "while updating dependencies for helm chart") && !conf.releaseConf.DryRun {
+				return true
+			}
+			return false
+		}),
+		retry.Attempts(60),
+		retry.Delay(1*time.Minute),
+		retry.OnRetry(func(n uint, err error) {
+			log.Printf("retry #%d: %s\n", n, err)
+		}),
+	)
+
+	return nil
+}
+
+func updateHelmRepositories() error {
+	// simulates helm repo update
+	f, err := repo.LoadFile(cli.New().RepositoryConfig)
+	if err != nil {
+		return fmt.Errorf("while loading repository file: %w", err)
+	}
+	for _, r := range f.Repositories {
+		cr, err := repo.NewChartRepository(r, getter.All(cli.New()))
+		if err != nil {
+			return fmt.Errorf("while getting new chart repository: %w", err)
+		}
+		if _, err = cr.DownloadIndexFile(); err != nil {
+			return fmt.Errorf("while updating repository index for (%s): %w", cr.Config.Name, err)
+		}
+		log.Printf("Updated repository index for (%s)\n", cr.Config.Name)
+	}
+	return nil
+}
+
+// uploadCharts will perform the following:
+//  1. Ensure that the Helm "stable" repository exists in the repository list.
+//  2. For each Helm chart, copy the chart to a temporary directory
+//  3. For each Helm chart's dependency, if the dependency's repository
+//     is set to 'https://helm.elastic.co', and the flag 'charts-repo-url'
+//     is set to another repository (such as a dev helm repo), the dependency's
+//     repository will be re-written to the given 'charts-repo-url' flag.
+//  4. Run 'Helm dependency update' for the Helm Chart.
+//  5. Run 'Helm package chart' for the Helm Chart to generate a tarball.
+//  6. Copy the tarball to the GCS bucket.
+func uploadCharts(ctx context.Context, tempDir string, charts []chart, conf ReleaseConfig) error {
+	type helmRepo struct {
+		name, url string
+	}
+	for _, r := range []helmRepo{{name: "stable", url: stableHelmChartsURL}, {name: conf.Bucket, url: conf.ChartsRepoURL}} {
```

Comment:
We add Helm repositories even if there are no charts to download? Must be harmless I guess?

Reply:
I've updated this logic to break early.  It's harmless though.